format/leveldb/leveldb_table.go

```diff
@@ -441,3 +446,21 @@ func mask(crc uint32) uint32 {
 	// Rotate right by 15 bits and add a constant.
 	return ((crc >> 15) | (crc << 17)) + kMaskDelta
 }
+
+// Concatinate byteslices and convert into a string.
+func stringify(byteSlices ...[]byte) string {
```

Comment:
If you want to save on allocations you could use a `strings.Builder` or possible sum all lengths allocate once with `make()` and then append. But i have feeling it wont make much difference in this case?

Reply:
![image](https://github.com/wader/fq/assets/216956/9854c535-6300-495c-84fa-446e321df438)

`strings.Builder` seems to do the same thing. I suppose the resizing would be made in an amortized-constant way (similar to hash tables)? I can still preallocate though. :)

Reply:
![image](https://github.com/wader/fq/assets/216956/bb3169f9-e01e-427c-9be9-d77fe0558f45)

Ok, this explains more what's going on. It seems sometimes it's mutated (and then I assume that's amortized constant)... and sometimes it's created from scratch (which is costly). It is not clear to me which one happens when though.

Reply:
Huh i assumed `strings.Builder` did more smart things by growing smarty but maybe the only smartness is in the `String()` method? but i do wonder if append might have some smartness about how to grow so it won't end up reallocating/copy all the time?

Reply:
@wader I'm trying to track down how append works precisely; this is closest I've found: https://github.com/golang/go/blob/master/src/cmd/compile/internal/ssagen/ssa.go#L3503
Don't have more clarity now. :)

Reply:
interesting! it seems to generate a call to growslice https://github.com/golang/go/blob/fe1b2f95e6dbfb6e6212bb391706ae62eb0ae5ec/src/runtime/slice.go#L155 which in turn seems to call nextslicecaphttps://github.com/golang/go/blob/fe1b2f95e6dbfb6e6212bb391706ae62eb0ae5ec/src/runtime/slice.go#L267 which has some heuristics how to grow

Reply:
https://godbolt.org/z/rEPerb3b4 maybe interesting. i would have assumed that append would just end up being a call to something but seem there lots of tricky going on, maybe be speed up common cases etc?

Reply:
@wader Oh, checking the compiler explorer was very insightful!! `runtime.growslice(SB)` clearly shows the connection to slice.go now. I don't quite see the big picture yet, but now I know where to start if I wanted to dive deeper. :)